File: Others/medium/StaircaseTraversal.py
def dp(height, max_steps, memo):

    if height == 0:
        return 1

    if memo[height] > 0:
        return memo[height]

    ways = 0
    for i in range(max(height - max_steps, 0), height):
        ways += dp(i, max_steps, memo)
    memo[height] = ways
    return ways


def staircaseTraversal(height, maxSteps):
    memo = [-1] * (height + 1)
    dp(height, maxSteps, memo)
    return memo[height]

File: Golang/medium/staircase-traversal.go
package main

func staircaseTraversal(height int, maxSteps int, memo map[int]int) int {
	if height < 0 {
		return 0
	}
	if height <= 1 {
		return 1
	}
	if val, ok := memo[height]; ok {
		return val
	}
	res := 0

	for idx := 1; idx <= maxSteps; idx++ {
		res += staircaseTraversal(height-idx, maxSteps, memo)
	}

	memo[height] = res
	return res
}

func StaircaseTraversal(height int, maxSteps int) int {
	return staircaseTraversal(height, maxSteps, map[int]int{})
}
